### ISSUE TITLE
Creating new file/folder in Explorer is aborted when files are concurrently being created/deleted

### ISSUE THREAD

**Issue by jaminthorns** [microsoft/vscode#112293]
### System Information

- VSCode Version: 1.52.0
- OS Version: macOS Catalina 10.15.7

### Steps to Reproduce

1. Have a process running that is creating/deleting files within a folder in the Explorer view.
2. Attempt to create a new file or folder within the Explorer view.
3. Observe that on creation/deletion of each file, the inline input for naming the new file/folder is exited and the creation process is aborted.

**Does this issue occur when all extensions are disabled?:** Yes

### Demonstration

![explorer_issue_demonstration](https://user-images.githubusercontent.com/6618434/101917147-86078780-3b8d-11eb-954f-1d75417748d6.gif)

### Notes

- This does not happen with file modifications, only creations and deletions.
- The issue occurs regardless of whether the folder in the Explorer view is collapsed, and it happens for folders that are fully collapsed and nested multiple layers down.
- I encountered this issue while installing dependencies for a project (a reasonably long-running task) and trying to create a new file through the Explorer view.

**Comment by isidorn** [comment]
Dup of https://github.com/microsoft/vscode/issues/111652
Fixed in latest vscode insiders

**Comment by jaminthorns** [comment]
@isidorn This is not a duplicate of #111652. That issue was related to unnecessary refreshing of the Explorer view immediately after adding a new folder/file (which creates a file change event) through the UI. The issue was fixed with 8b9b8df0247d637d2a5216e96a2096cb109a6288 by not refreshing the view for a file change event if the file/folder was added through the UI.

This issue has the same root cause: the Explorer view is refreshing and interrupting the flow of creating a new file/folder by discarding the state of the input. This situation is different, though, in that the file/folder being created is not through the UI, it's through a background process. The fix for #111652 does not address this.

I've verified that this issue is _not_ fixed on the latest Insiders build:

    Version: 1.52.0-insider
    Commit: dbbf7079b20b3036b3ff5ac2e2106b84dd3c6d66
    Date: 2020-12-10T16:23:20.433Z
    Electron: 9.3.5
    Chrome: 83.0.4103.122
    Node.js: 12.14.1
    V8: 8.3.110.13-electron.0
    OS: Darwin x64 19.6.0

I see 2 possible solutions:

1. Pause all refreshing of the Explorer view while creating a file/folder. If a file change event comes through while creating a file/folder, queue it until that the creation flow is finished (if multiple events come through, you only need to refresh once after the file/folder is created). This would mean that the Explorer view is "stale" when creating a file/folder.

2. Refresh the Explorer view as the creation flow is happening while somehow maintaining the state of the creation flow (file/folder destination and name). This would ensure the that Explorer view stays "live", but it could mean that the input for file/folder creation jumps around.

I personally think that solution 1 would be more user friendly and (after looking at 8b9b8df0247d637d2a5216e96a2096cb109a6288) probably easier to implement. The "staleness" of the Explorer view when creating a file/folder wouldn't be unexpected, since every part of the view fades when that input is focused already.

Also, I'm not sure what the behavior was in VS Code 1.51. This might have always been an issue beforehand that I never noticed, or it might be the recent update to 1.52.

**Comment by isidorn** [comment]
Can you please retry with latest insiders 76436a4d43566cd313629799e2ec51b4d00caa64

**Comment by jaminthorns** [comment]
Just tried it with the latest Insiders build:

```
Version: 1.53.0-insider
Commit: 96b426ef1da0f0a51ce9ef1931cd1fd2322547e4
Date: 2020-12-14T05:49:45.285Z
Electron: 11.0.3
Chrome: 87.0.4280.67
Node.js: 12.18.3
V8: 8.7.220.25-electron.0
OS: Darwin x64 19.6.0
```

The issue is still present (same behavior as the video in the top of this screenshot).

**Comment by isidorn** [comment]
Thanks for trying!
Let's reopen. However I believe we were always behaving like this, the issue is that when there are new file events we have to refresh the tree so those changes would show up. I guess we should somehow fix in a way that a tree refresh would preserve the input box.

**Comment by ducalpha** [comment]
I encounter a similar issue: When files in the tree are being created (e.g., data files are generated), the keyboard cursor focus is continuously stolen by the Explorer. Even my typing on editor is interrupted and the focus changes to the Explorer. My current workaround is to close the Explorer panel. This is a remote project connected via ssh. 

However, on another VS Code on another machine connecting to the same project via ssh, there is no such issue. Maybe it is due to some setting.

**Comment by isidorn** [comment]
@ducalpha please file a new issue if you can reproduce the problem with latest vscode insdiers
https://code.visualstudio.com/insiders/

**Comment by cloudhan** [comment]
#117116 should be the same issue as this one. In that issue, it is git extension that triggers the refreshing. And it seems passively refresh interactable UI element is a bad idea XD. My take on this, it need something like queued message handling mechanism, and let the interactable part decide when to properly consume an event, otherwise, these type of interruption might scatter over all system. Just my two cents...

**Comment by yepitschunked** [comment]
I have the same problem as @cloudhan and can confirm that it's resolved if I disable the Git integration (`"git.enabled": false` in your settings json)

**Comment by Silic0nS0ldier** [comment]
Git seems to be the cause of the issue for me, appearing to have formed some kind of loop whose termination condition I have yet to identify.

```
(loop start)
> git status -z -u
> git ls-files --stage -- /path-to-active-workspce/path-to-last-changed-file
> git show --textconv :path-to-last-changed-file
> git symbolic-ref --short HEAD
> git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/current-branch-name refs/remotes/current-branch-name
> git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname)
> git remote --verbose
> git ls-files --stage -- /path-to-active-workspce/path-to-last-changed-file
> git show --textconv :path-to-last-changed-file
> git config --get commit.template
> git ls-files --stage -- /path-to-active-workspce/path-to-last-changed-file
> git show --textconv :path-to-last-changed-file
> git ls-files --stage -- /path-to-active-workspce/path-to-last-changed-file
> git show --textconv :path-to-last-changed-file
> git ls-files --stage -- /path-to-active-workspce/path-to-last-changed-file
> git show --textconv :path-to-last-changed-file
(loop end)
```

Fairly clear something is going haywire with git, but that just distracts from the real issue (user input interruption).

The editor has to deal with similar problems (file events happening in parallel to user input) and thankfully deals with the challenge a lot better. If a file with unsaved changes (or is open) is deleted, the file is preserved with `(deleted)` added to the tab. On save, the user is given the choice of what to do. A similar workflow occurs when a file is modified in the background.

Taking a page from the editor, I would expect the following behaviour from the file explorer. If changes occur to the tree, preserve current user input at all costs. When it is time to persist the change, deal with the change appropriately (applying change if "safe" to do so, prompting for a resolution if there is a change). Changes to the tree can occur, input must be preserved (meaning changes may occur "around" the user input).

**Comment by yepitschunked** [comment]
I noticed that #126793 mentioned using the watchman fsmonitor, which I'm also using. Disabling it (via `git config --unset core.fsmonitor`) fixed the issue for me. 

I suspect a related issue is mentioned in this SO question: https://stackoverflow.com/questions/58006273/watchman-makes-fsnotify-spuriously-detect-file-changes

There is a watcher for the .git subdirectory: https://github.com/microsoft/vscode/blob/94c9ea46838a9a619aeafb7e8afd1170c967bb55/extensions/git/src/watch.ts#L16

This causes a loop because most git status will issue a watchman query, which pokes the watcher, which issues another git status, ad infinitum.

**Comment by Silic0nS0ldier** [comment]
It won't address _all_ scenarios that interrupt file creation/renaming, but it'll likely address the vast majority of them in my case. Thanks for the investigation and PR with proposed workaround.

**Comment by yepitschunked** [comment]
@JacksonKearl can you take a look at https://github.com/microsoft/vscode/issues/112293#issuecomment-871666315 and the PR I submitted?

**Comment by JacksonKearl** [comment]
I'm not sold on the idea of incorporating a bunch of filters in core to try to pre-empt what third party software may or may not be doing. But cc @eamodio as the owner of the git area.

For this ticket I believe the best solution is to just pause reacting to watch events at all while the editing input is active.

**Comment by JacksonKearl** [comment]
We now pause updates to the explorer when an edit is underway. This should resolve the annoyances while being much simpler than trying to update the explorer in the background while preserving the state of the edit.

**Comment by LukeWolfGames** [comment]
> We now pause updates to the explorer when an edit is underway. This should resolve the annoyances while being much simpler than trying to update the explorer in the background while preserving the state of the edit.

This idea sounds more practical. Is a solution not released yet?

**Comment by JacksonKearl** [comment]
It's out in Insiders as of mid December, but as we don't have a stable December release it is not out in stable

**Comment by LukeWolfGames** [comment]
I understand. Thanks!

**Comment by rchiodo** [comment]
/verified

File Explorer doesn't update while it's creating a file.
